drivers: ieee802154: SSED attachment management

xavraz opened the pull request:
Few functions have been corrected to the SSED attachment for the MCXW72 as leader :
- mcxw_tx(): tx_frame.tx_delay corrected. Adjusts the scheduled transmission start time to account for the 192µs PHY preamble/header overhead before the MAC payload, converting to symbol units, or defaults to immediate transmission if the time is too soon.

- mcxw_rx_thread(): Stores the frame counter and key ID from a security-enabled enhanced ACK into the received packet's metadata for cryptographic verification.

- mcxw_tx_started() added: Notifies the registered event handler that a frame transmission has started, passing the frame buffer as context.

- pd_mac_sap_handler(): timestamp corrected

- mcxw_configure(): IEEE802154_CONFIG_EVENT_HANDLER case updated

xavraz committed:
drivers: ieee802154: SSED attachment management

Few functions have been corrected to the SSED attachment for the
MCXW72 as leader :
- mcxw_tx():
  tx_frame.tx_delay corrected.
  Adjusts the scheduled transmission start time to account for the SHR
  duration (160us) as requested by the OpenThread radio API.

- mcxw_rx_thread():
  Stores the frame counter and key ID from a
  security-enabled enhanced ACK into the received packet's metadata
  for cryptographic verification.

- mcxw_tx_started() added:
  Notifies the registered event handler that a frame transmission has
  started, passing the frame buffer as context.

- pd_mac_sap_handler(): timestamp corrected

- mcxw_configure(): IEEE802154_CONFIG_EVENT_HANDLER case updated

Signed-off-by: Xavier Razavet <xavier.razavet@nxp.com>

rlubos commented on samples/net/openthread/shell/prj-ot-host.conf:
```diff
@@ -55,3 +51,8 @@ CONFIG_OPENTHREAD_SRP_CLIENT=y
 CONFIG_OPENTHREAD_IP6_FRAGM=y
+
+CONFIG_OPENTHREAD_PLATFORM_CSL_UNCERT=20
+CONFIG_IEEE802154_MCXW_CSL_ACCURACY=50
```
Driver-specific configs should not be placed in a generic config file,  please put it in board specific config in `boards` subdirectory.

rlubos commented:
I don't think assignees were set correctly given it's a change in NXP driver.

axelnxp commented on samples/net/openthread/shell/prj-ot-host.conf:
```diff
@@ -55,3 +51,8 @@ CONFIG_OPENTHREAD_SRP_CLIENT=y
+CONFIG_OPENTHREAD_PLATFORM_CSL_UNCERT=20
+CONFIG_IEEE802154_MCXW_CSL_ACCURACY=50
+CONFIG_NET_PKT_TXTIME=y
+CONFIG_NET_PKT_TIMESTAMP=y
```
My understanding is that those are required by the driver, am I right ?
If so, make it explicit in the driver kconfig.
Either use `select` (or `imply` if it's optional, meaning the driver can build without it, which I doubt anyway) or `depends on` (if you want to put the reponsibility to the application to enable it).
I'd go for `select` in this case.

JarEmb commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -540,10 +552,22 @@ static int mcxw_tx(const struct device *dev, enum ieee802154_tx_mode mode, struc
 			rf_adjust_tstamp_from_app(mcxw_radio->tx_frame.tx_delay);
-		msg->msgData.dataReq.startTime /= IEEE802154_SYMBOL_TIME_US;
+
+		if ((msg->msgData.dataReq.startTime > 192) &&
```
it is thread specific, I agree to put it in radio.c

rlubos replied:
Sorry I'm not up to date with Thread 1.4 specs, but looking at the OpenThread documentation:
https://github.com/zephyrproject-rtos/openthread/blob/2bc7712f57af22058770d1ef131ad3da79a0c764/include/openthread/platform/radio.h#L282-L284
They claim to set the timestamp to the end of SFD field (so start of PHR as I wrote above), and in OpenThread radio layer:
https://github.com/zephyrproject-rtos/zephyr/blob/ab09c94377c562deb7f1b1feca1fce55e1819e02/modules/openthread/platform/radio.c#L419-L423
We just pass the value as-is to the driver.

So the timestamp the driver receives should be the start of PHR.

xavraz committed:
samples: openthread: shell: .conf files updated

frdm_mcxw71.conf: LOGs removed for RAM size optimization.
prj-ot-host.conf: LOGs removed because already in prj.conf.

Signed-off-by: Xavier Razavet <xavier.razavet@nxp.com>

xavraz committed:
dts: arm: nxp: MCXW72 corrected on the lptmr clock frequency

The lptmr clock frequency was wrong and equal to 32000.
The clock frequency is 32.768KHz (32768).

Signed-off-by: Xavier Razavet <xavier.razavet@nxp.com>

xavraz committed:
drivers: ieee802154: mcxw ieee802154 driver updated

To support the SSED attachment, the IEEE802154_MCXW_CSL_ACCURACY
default value shall be 50.

Signed-off-by: Xavier Razavet <xavier.razavet@nxp.com>

xavraz committed:
soc: nxp:  MCXW7x configuration updated for CSL

The default MCXW7x platform's clock uncertainty in microseconds for
Coordinated Sampled Listening (CSL) timing synchronization in
OpenThread shall be set to 20.

Signed-off-by: Xavier Razavet <xavier.razavet@nxp.com>

axelnxp commented on samples/net/openthread/shell/prj-ot-host.conf:
```diff
@@ -5,8 +5,8 @@ CONFIG_NETWORKING=y
-CONFIG_LOG=y
-CONFIG_NET_LOG=y
+CONFIG_LOG=n
+CONFIG_NET_LOG=n
```
If you really want to disable logs by default in the app, do it at board level (create a <board>.conf in the `boards` subfolder) and not in the common configuration file

George-Stefan commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -540,10 +549,21 @@ static int mcxw_tx(const struct device *dev, enum ieee802154_tx_mode mode, struc
+		/* Compensates the 192µs SHR duration by adjusting the MAC payload start
+		 * time backwards and converting it to symbol periods.
+		 */
+		if ((msg->msgData.dataReq.startTime > IEEE802154_SHR_DURATION_US) &&
```
start time expiration is done on the NBU side by the PHY.
There is no need to do any validation here

George-Stefan commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -600,16 +620,13 @@ static int mcxw_tx(const struct device *dev, enum ieee802154_tx_mode mode, struc
 	if (phy_status == gPhySuccess_c) {
 		mcxw_radio->tx_status = 0;
 		mcxw_radio->state = RADIO_STATE_TRANSMIT;
```
move mcxw_radio->state = RADIO_STATE_TRANSMIT; before MAC_PD_SapHandler() to avoid a race condition with tx aborted event. Don't update the state if tx request (MAC_PD_SapHandler()) fails.
For rx, it's correct https://github.com/nxp-upstream/zephyr/blob/ecfda026b2b74885c68e1b279badce7b80fdc7b1/drivers/ieee802154/ieee802154_mcxw.c#L308

xavraz commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -540,10 +549,21 @@ static int mcxw_tx(const struct device *dev, enum ieee802154_tx_mode mode, struc
+		/* Compensates the 192µs SHR duration by adjusting the MAC payload start
+		 * time backwards and converting it to symbol periods.
+		 */
+		if ((msg->msgData.dataReq.startTime > IEEE802154_SHR_DURATION_US) &&
```
Hello George,
I have removed the test.

George-Stefan commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -1005,8 +1015,8 @@ static int mcxw_set_channel(const struct device *dev, uint16_t channel)
 {
-	static uint64_t sw_timestamp;
-	static uint64_t hw_timestamp;
+	static uint64_t sw_timestamp; /* µs, on the last timestamp, monotonous */
```
typo: `on the last timestamp` should be `the last timestamp`

xavraz replied:
Corrected

George-Stefan reviewed (changes requested):
Please update the 3rd commit message:
```
[dts: arm: nxp: MCXW72 corrected on the lptmr clock frequency]
The lptmr clock frequency was wrong and equal to 32000.
The clock frequency is 32MHz (32768).
```
`32MHz` should be `32.768KHz`

Please update the last commit message:
```
tx_frame.tx_delay corrected.
  Adjusts the scheduled transmission start time to account for the
  192µs PHY preamble/header overhead before the MAC payload,
  converting to symbol units, or defaults to immediate transmission
  if the time is too soon.
```

should be
```
Adjusts the scheduled transmission start time to account for the SHR
duration (160us) as requested by the OpenThread radio API
```

George-Stefan commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -540,10 +549,11 @@ static int mcxw_tx(const struct device *dev, enum ieee802154_tx_mode mode, struc
 	case IEEE802154_TX_MODE_TXTIME:
 	case IEEE802154_TX_MODE_TXTIME_CCA:
-		mcxw_radio->tx_frame.tx_delay = net_pkt_timestamp_ns(pkt);
+		mcxw_radio->tx_frame.tx_delay -= IEEE802154_SHR_DURATION_US;
```
This is not correct,
`tx_delay` is in nanoseconds while SHR duration is in us.

See https://github.com/zephyrproject-rtos/zephyr/pull/100389/files#r2601444923
